### DISCUSSION TITLE
몇가지 수정사항

### DISCUSSION THREAD

**Discussion by sonarmans** [oleneyl/maplestory_dpm_calc#664] [General]
1. 방무 필수 직업군에 대한 미고려: 유니온 8000 기준으로도 무보엠 공퍼 대신 방무 한줄이 필수적으로 들어가야 하는 직업들이 있음.

2. 호영 하이퍼스킬 설정 오류: 실전에서 사용하는 하이퍼스킬들과 다름

3. 호영 극딜 준비시간 미고려: 선단 먹은 후 이중게이지로 극딜 준비하는 시간에 대한 고려가 덜 됨

**Comment by icepeng** [comment]
1. 현재 스펙 최적화는 `JobGenerator.package()` 메소드 내부에서 하이퍼스탯 -> 유니온배치 -> 무보엠 순으로 진행하고 있으며, 이 과정에서 방무의 가치가 보공보다 높은 경우 방무가 선택됩니다.

2. 하이퍼스킬은 실전 사용 여부보다 dpm을 우선시하고 있습니다만, 이 부분은 기준을 어떻게 하는게 좋을지 잘 모르겠습니다. 우선 말씀해주셨던 이그노어 가드 2개 투자하는것이 더 좋은지 확인해보겠습니다.

3. 어떤 부분을 보고 그렇게 판단하셨는지 잘 모르겠습니다. 게이지는 시뮬레이션 위에 구현되어 있으며, 게이지가 부족하면 스킬이 실행되지 않으므로, 극딜을 약간 비효율적으로 사용하고 있을지는 모르나 불가능한 시간 내에 준비를 끝마치는 일은 없습니다.

```py
        # 부적 도술
        TalismanConstraint = core.ConstraintElement("부적 도력 100", TalismanEnergy, partial(TalismanEnergy.judge, 100, 1))
        TalismanConsume = TalismanEnergy.stackController(-100)
        TalismanConsume.onAfter(gainEnergy(ScrollEnergy, 200))
        for sk in [Mabong, Talisman_Clone, Talisman_Seeker]:
            sk.onConstraint(TalismanConstraint)
            sk.onAfter(TalismanConsume)

        # 두루마리 도술
        ScrollConstraint = core.ConstraintElement("두루마리 도력 900", ScrollEnergy, partial(ScrollEnergy.judge, 900, 1))
        ScrollConsume = ScrollEnergy.stackController(-900)
        for sk in [Misaeng, Waryu, Butterfly_Dream, Summon_Sanryung]:
            sk.onConstraint(ScrollConstraint)
            sk.onAfter(ScrollConsume)
```

관련 코드입니다. TalismanConstraint, ScrollConstraint가 게이지가 부족할 때 스킬의 실행을 제한하는 역할을 합니다.

**Comment by sonarmans** [comment]
1. 확인했습니다

2. 이 부분이 고려된 것이라고 생각을 못하고 코멘트를 적었는데, 고려된 것이라면 기존 방무 수치도 전부 달라질 수 있겠네요. 

3. 코스트직업이기 때문에 사람이 누르는것과 코드로 구현한것을 실제로 누름에 있어서 차이가 크다는 것을 지적한것입니다. 펫 자동 스킬 처럼 모든 극딜 준비 스킬들이 코스트임에도 자동으로 스킬이 구현되는 특성상 실제 극딜 준비보다 훨씬 짧은 시간에 극딜 준비가 이뤄집니다. 극딜준비에서 코드구현 타임라인과 실제 유저의 허수아비 딜링에서의 극딜 준비 시간을 비교해보시면 좋을 것 같습니다. 호영은 원래 타 직업 대비 극딜 준비 시간이 깁니다. 그것이 코드로 구현되었을때 덜 고려됨을 지적한것입니다. 이는 에반에게도 동일하게 실제 딜링시 조디악 발동보다 코드 구현시 조디악 발동이 훨씬 빠르기 때문에 실제 dpm보다 코드상의 dpm이 훨씬 잘 나오게 됩니다. 극딜 준비 시간을 상수로 처리해서 딜링 사이클을 만드시는게 더 적합하다고 봅니다.

**Reply by icepeng** [reply]
극딜 준비시간을 상수화 하는것은 어렵기도 하고, 애매하기도 합니다.

버프와 버프가 연속으로 사용될 때, 사이에 일정 딜레이 (30~90ms)가 삽입되도록 하는것이 해결책에 가깝다고 봅니다.

**Reply by sonarmans** [reply]
아, 잘못이해했습니다. 글 지울게요

**Reply by icepeng** [reply]
버프의 고유 딜레이는 이미 고려되어 있습니다.

**Reply by icepeng** [reply]
예시) 신궁 불스아이

```py
        BoolsEye = core.BuffSkill(
            name="불스아이",
            delay=960,
            remain=30 * 1000,
            cooltime=90 * 1000,
            crit=20,
            crit_damage=10,
            armor_ignore=20,
            pdamage=20,
        ).wrap(core.BuffSkillWrapper)
```

**Reply by sonarmans** [reply]
네 고려된걸 깜빡하고 글 썼다가 지웠습니다

**Comment by icepeng** [comment]
1 -> 답변 완료
2 -> 호영 하이퍼스킬 한번 더 확인해볼 것
3 -> 답변 완료

정리가 되었다고 생각되어, 답변을 마치겠습니다. 감사합니다.
